src/java_bytecode/jar_file.cpp

```diff
@@ -1,131 +1,86 @@
 {
-  if(mz_ok)
+  const auto entry=m_name_to_id.find(name);
+  INVARIANT(entry!=m_name_to_id.end(), "File doesn't exist");
+  try
+  {
+    return m_zip_archive.extract(entry->second);
+  }
+  catch(const std::runtime_error &)
   {
-    mz_zip_reader_end(&zip);
-    mz_ok=false;
+    return "";
   }
 }
 
-std::string jar_filet::get_entry(const irep_idt &name)
+/// Remove leading and trailing whitespace characters from string
+static std::string trim(
+  std::string::const_iterator begin,
+  std::string::const_iterator end)
 {
-  if(!mz_ok)
-    return std::string("");
-
-  std::string dest;
-
-  auto entry=filtered_jar.find(name);
-  assert(entry!=filtered_jar.end());
-
-  size_t real_index=entry->second;
-  mz_zip_archive_file_stat file_stat;
-  memset(&file_stat, 0, sizeof(file_stat));
-  mz_bool stat_ok=mz_zip_reader_file_stat(&zip, real_index, &file_stat);
-  if(stat_ok!=MZ_TRUE)
-    return std::string();
-  std::vector<char> buffer;
-  size_t bufsize=file_stat.m_uncomp_size;
-  buffer.resize(bufsize);
-  mz_bool read_ok=
-    mz_zip_reader_extract_to_mem(&zip, real_index, buffer.data(), bufsize, 0);
-  if(read_ok!=MZ_TRUE)
-    return std::string();
-
-  dest.insert(dest.end(), buffer.begin(), buffer.end());
-
-  return dest;
+  while(begin!=end && std::isspace(*begin))
+    ++begin;
+  if(begin!=end)
+  {
+    --end;
+    while(begin!=end && std::isspace(*end))
+      --end;
+    end++;
+  }
+  return { begin, end };
 }
 
-jar_filet::manifestt jar_filet::get_manifest()
+std::unordered_map<std::string, std::string> jar_filet::get_manifest()
 {
-  auto entry=filtered_jar.find("META-INF/MANIFEST.MF");
-  if(entry==filtered_jar.end())
-    return manifestt();
-
-  std::string dest=get_entry(entry->first);
-  std::istringstream in(dest);
-
-  manifestt manifest;
-
-  std::string line;
-  while(std::getline(in, line))
+  std::unordered_map<std::string, std::string> out;
+  const auto entry=m_name_to_id.find("META-INF/MANIFEST.MF");
+  if(entry!=m_name_to_id.end())
   {
-    std::size_t pos=line.find(':');
-    if(pos==std::string::npos)
-      continue;
-    std::string key=line.substr(0, pos);
-
-    // skip spaces
-    pos++;
-    while(pos<line.size() && line[pos]==' ') pos++;
-
-    std::string value=line.substr(pos, std::string::npos);
-
-    // trim off \r
-    if(!value.empty() && *value.rbegin()=='\r')
-      value.resize(value.size()-1);
-
-    // store
-    manifest[key]=value;
+    std::istringstream in(this->get_entry(entry->first));
+    std::string line;
+    while(std::getline(in, line))
+    {
+      const auto key_end=std::find(line.cbegin(), line.cend(), ':');
+      if(key_end!=line.cend())
+        out.emplace(
+          trim(line.cbegin(), key_end),
+          trim(std::next(key_end), line.cend()));
+    }
   }
-
-  return manifest;
+  return out;
 }
+
+std::vector<std::string> jar_filet::filenames() const
+{
+  std::vector<std::string> out;
+  for(const auto &pair : m_name_to_id)
+    out.emplace_back(pair.first);
```

Comment:
It would be cheaper to use range-insert - I wish we had an iterator adapter library...